src/debugging-support-in-rustc.md

```diff
@@ -0,0 +1,259 @@
+
+1. Enum: Needed for support for enum types. The Rustc writes the information about enum into 
+the DWARF and GDB reads the DWARF to understand where is the tag field or is there a tag 
+field or is the tag slot shared with non-zero optimization etc.
+
+2. Dissect trait objects: DWARF extension where the trait object's description in the DWARF 
+also points to a stub description of the corresponding vtable which in turn points to the 
+concrete type for which this trait object exists. This means that you can do a `print *object` 
+for that trait object, and GDB will understand how to find the correct type of the payload in 
+the trait object.
+
+#### Developer notes
+
+* This work is now upstream. Bugs can be reported in [GDB Bugzilla].
+
+### LLDB
+
+We have our own fork of LLDB - [https://github.com/rust-lang/lldb]
+Fork of LLVM project - [https://github.com/rust-lang/llvm-project]
+
+#### Rust expression parser
+
+This expression parser is written in C++. It is a type of [Recursive Descent parser]. 
+Implements slightly less of the Rust language than GDB. 
+[TODO: CHECK: if still true, Tom says he missed a step. he did not impl vtable]. 
+LLDB has Rust like value and type output.
+
+#### Parser extensions
+
+There is some special code in the DWARF reader in LLDB to support the extensions. 
+A couple of examples of DWARF reader support needed are as follows - 
+
+1. Enum: Needed for support for enum types. The Rustc writes the information about 
+enum into the DWARF and LLDB reads the DWARF to understand where is the tag field or 
+is there a tag field or is the tag slot shared with non-zero optimization etc. 
+In other words, it has enum support as well.
+
+#### Developer notes
+
+* None of the LLDB work is upstream. This [rust-lang/lldb wiki page] explains a few details. 
+* The reason for forking LLDB is that LLDB recently removed all the other language plugins 
+due to lack of maintenance.
+* LLDB has a plugin architecture but that does not work for language support. 
+* This is built on macOS because it was easier to solve. 
+* Tom has a possible solution which can enable us to ship LLDB everywhere. 
+* GDB generally works better on Linux.
+
+## DWARF and Rustc
+
+[DWARF] is the standard way compilers generate debugging information that debuggers read. 
+It is _the_ debugging format on macOS and Linux. It is a multi-language, extensible format 
+and is mostly good enough for Rust's purposes. Hence, the current implementation reuses DWARF's 
+concepts. This is true even if some of the concepts in DWARF do not align with Rust
+semantically because generally there can be some kind of mapping between the two.
+
+We have some DWARF extensions that the Rust compiler emits and the debuggers understand that 
+are _not_ in the DWARF standard.
+
+* vtable thing
+* .... Does not remember...
+[TODO: Ask Tom for this? they are documented in debuggers in comments or LLVM in comments]
```

Comment:
@tromey Can you point out the location of these extensions please?

Reply:
They are only documented as comments in the LLVM (and maybe gdb and lldb) source.

The vtable thing is that the Rust compiler will emit DWARF for a virtual table, and this vtable object will have a `DW_AT_containing_type` that points to the real type.  This lets debuggers dissect a trait object pointer to correctly find the payload.  E.g., here's such a DIE, from a test case in the gdb repository:

```
 <1><1a9>: Abbrev Number: 3 (DW_TAG_structure_type)
    <1aa>   DW_AT_containing_type: <0x1b4>
    <1ae>   DW_AT_name        : (indirect string, offset: 0x23d): vtable
    <1b2>   DW_AT_byte_size   : 0
    <1b3>   DW_AT_alignment   : 8
```

The other extension that I remember is that the Rust compiler can emit a tagless discriminated union.  I filed a DWARF feature request for this: http://dwarfstd.org/ShowIssue.php?issue=180517.2